<a href="https://colab.research.google.com/github/ronaldnuunes/curso_senac_ia/blob/main/exercicios_consumindo_api_dummyjson.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercícios Práticos: Consumindo uma API REST com Python

## API escolhida: DummyJSON

Neste notebook, você irá praticar o consumo de uma API pública chamada **DummyJSON**.

Ela fornece dados fictícios em formato JSON, como produtos, usuários, carrinhos, posts e tarefas. É uma API muito útil para treinar consumo de API REST, análise de JSON e transformação de dados com Python.

### Documentação oficial

Acesse a documentação da API:

https://dummyjson.com/docs

Documentação específica de produtos:

https://dummyjson.com/docs/products

---

## Objetivos dos exercícios

Ao final deste notebook, você deverá ser capaz de:

- Fazer requisições HTTP usando Python;
- Consumir endpoints com o método `GET`;
- Enviar dados usando `POST`;
- Atualizar dados usando `PUT` ou `PATCH`;
- Simular exclusão usando `DELETE`;
- Interpretar respostas em JSON;
- Transformar dados da API em tabelas com Pandas;
- Criar pequenas análises a partir dos dados recebidos.

---

## Importante

A DummyJSON é uma API de testes. As operações de criação, atualização e exclusão são **simuladas**.

Isso significa que a API responde como se tivesse criado, atualizado ou excluído um recurso, mas os dados não são realmente persistidos no servidor.

---

## Como usar este notebook

Leia cada enunciado com atenção e complete os trechos marcados com `TODO`.

Não apague as células de enunciado.

Quando terminar, execute todas as células na ordem.

# Parte 1 — Preparação do ambiente

Antes de consumir uma API, precisamos importar algumas bibliotecas.

Usaremos:

- `requests`: para fazer chamadas HTTP;
- `pandas`: para transformar dados JSON em tabelas;
- `json`: para exibir algumas respostas de forma mais organizada.

In [1]:
import requests
import pandas as pd
import json

Vamos definir a URL base da API.

Sempre que formos chamar um endpoint da DummyJSON, começaremos por essa URL.

In [2]:
BASE_URL = "https://dummyjson.com"

# Parte 2 — Primeiro contato com a API

## Exercício 1 — Testar se a API está funcionando

Faça uma requisição `GET` para o endpoint abaixo:

```text
https://dummyjson.com/test
```

Depois, exiba:

- O código de status HTTP;
- A resposta em JSON.

### Dica

Use:

```python
requests.get(...)
```

In [3]:
url = BASE_URL + "/test"

resposta = requests.get(url)

print("Status Code:", resposta.status_code)
print(resposta.json())

Status Code: 200
{'status': 'ok', 'method': 'GET'}


## Perguntas do Exercício 1

Responda abaixo, em texto:

1. Qual foi o status code retornado? 200
2. O que esse status code significa? Que a requisição foi realizada com sucesso.
3. A API respondeu em qual formato de dados? JSON

**Respostas:**

1. 200
2.Significa que a requisição foi realizada com sucesso.
3. JSON.

# Parte 3 — Consumindo produtos com GET

A partir de agora, trabalharemos com o recurso `products`.

Um recurso, em uma API REST, representa uma entidade do sistema.

Neste caso:

```text
/products
```

representa a coleção de produtos da API.

## Exercício 2 — Buscar todos os produtos

Faça uma requisição `GET` para:

```text
https://dummyjson.com/products
```

Depois, exiba:

- O status code;
- As chaves principais do JSON retornado;
- A quantidade de produtos retornados nesta primeira página;
- O total de produtos disponíveis na API.

### Observação

A resposta da API não é uma lista diretamente. Ela retorna um objeto JSON com algumas chaves, como:

- `products`
- `total`
- `skip`
- `limit`

In [4]:
url = BASE_URL + "/products"

resposta = requests.get(url)
dados = resposta.json()

print("Status Code:", resposta.status_code)
print("Chaves:", dados.keys())
print("Produtos retornados:", len(dados["products"]))
print("Total disponível:", dados["total"])

Status Code: 200
Chaves: dict_keys(['products', 'total', 'skip', 'limit'])
Produtos retornados: 30
Total disponível: 194


## Exercício 3 — Visualizar o primeiro produto

Usando os dados do exercício anterior, exiba apenas o primeiro produto da lista.

Depois, responda:

1. Qual é o nome do produto?
2. Qual é o preço?
3. Qual é a categoria?
4. Qual é a avaliação média, campo `rating`?

In [5]:
primeiro_produto = dados["products"][0]

print(json.dumps(primeiro_produto, indent=2, ensure_ascii=False))

{
  "id": 1,
  "title": "Essence Mascara Lash Princess",
  "description": "The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.",
  "category": "beauty",
  "price": 9.99,
  "discountPercentage": 10.48,
  "rating": 2.56,
  "stock": 99,
  "tags": [
    "beauty",
    "mascara"
  ],
  "brand": "Essence",
  "sku": "BEA-ESS-ESS-001",
  "weight": 4,
  "dimensions": {
    "width": 15.14,
    "height": 13.08,
    "depth": 22.99
  },
  "warrantyInformation": "1 week warranty",
  "shippingInformation": "Ships in 3-5 business days",
  "availabilityStatus": "In Stock",
  "reviews": [
    {
      "rating": 3,
      "comment": "Would not recommend!",
      "date": "2025-04-30T09:41:02.053Z",
      "reviewerName": "Eleanor Collins",
      "reviewerEmail": "eleanor.collins@x.dummyjson.com"
    },
    {
      "rating": 4,
      "comment": "Very satisfied!",
      "date": "2025-04-

**Respostas:**

1. Nome do produto: essence mascra lash princess
2. Preço: 9.99
3. Categoria: beauty
4. Rating: 2.56

# Parte 4 — Transformando JSON em tabela

Em projetos de dados e IA, consumir uma API normalmente é apenas o começo.

Depois de receber os dados em JSON, é comum transformar esses dados em uma tabela para análise.

## Exercício 4 — Criar um DataFrame de produtos

Transforme a lista de produtos em um DataFrame do Pandas.

Depois, exiba as 5 primeiras linhas.

In [6]:
df_produtos = pd.DataFrame(dados["products"])

print(df_produtos.head())

   id                          title  \
0   1  Essence Mascara Lash Princess   
1   2  Eyeshadow Palette with Mirror   
2   3                Powder Canister   
3   4                   Red Lipstick   
4   5                Red Nail Polish   

                                         description category  price  \
0  The Essence Mascara Lash Princess is a popular...   beauty   9.99   
1  The Eyeshadow Palette with Mirror offers a ver...   beauty  19.99   
2  The Powder Canister is a finely milled setting...   beauty  14.99   
3  The Red Lipstick is a classic and bold choice ...   beauty  12.99   
4  The Red Nail Polish offers a rich and glossy r...   beauty   8.99   

   discountPercentage  rating  stock                   tags           brand  \
0               10.48    2.56     99      [beauty, mascara]         Essence   
1               18.19    2.86     34    [beauty, eyeshadow]  Glamour Beauty   
2                9.84    4.64     89  [beauty, face powder]    Velvet Touch   
3         

## Exercício 5 — Inspecionar o DataFrame

Agora que você tem uma tabela, descubra:

1. Quantas linhas e colunas existem?
2. Quais são os nomes das colunas?
3. Quais colunas parecem numéricas?
4. Quais colunas parecem textuais?

In [11]:
print(df_produtos.shape[1])

print(df_produtos.columns)

print(df_produtos.info())

22
Index(['id', 'title', 'description', 'category', 'price', 'discountPercentage',
       'rating', 'stock', 'tags', 'brand', 'sku', 'weight', 'dimensions',
       'warrantyInformation', 'shippingInformation', 'availabilityStatus',
       'reviews', 'returnPolicy', 'minimumOrderQuantity', 'meta', 'images',
       'thumbnail'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    30 non-null     int64  
 1   title                 30 non-null     object 
 2   description           30 non-null     object 
 3   category              30 non-null     object 
 4   price                 30 non-null     float64
 5   discountPercentage    30 non-null     float64
 6   rating                30 non-null     float64
 7   stock                 30 non-null     int64  
 8   tags                  30 non-nu

**Respostas:**

1. Linhas e colunas: 30 linhas e 22 colunas.
2. Colunas existentes:
3. Colunas numéricas:
4. Colunas textuais:

In [12]:
# Colunas numéricas
print(df_produtos.select_dtypes(include=['int64', 'float64']).columns)

# Colunas textuais
print(df_produtos.select_dtypes(include=['object']).columns)

Index(['id', 'price', 'discountPercentage', 'rating', 'stock', 'weight',
       'minimumOrderQuantity'],
      dtype='object')
Index(['title', 'description', 'category', 'tags', 'brand', 'sku',
       'dimensions', 'warrantyInformation', 'shippingInformation',
       'availabilityStatus', 'reviews', 'returnPolicy', 'meta', 'images',
       'thumbnail'],
      dtype='object')


# Parte 5 — GET com parâmetros

Muitas APIs permitem enviar parâmetros na URL.

Exemplo:

```text
https://dummyjson.com/products?limit=10&skip=0
```

Nessa URL:

- `limit=10` indica que queremos no máximo 10 produtos;
- `skip=0` indica que não queremos pular nenhum produto.

Em Python, podemos enviar esses parâmetros usando o argumento `params` da biblioteca `requests`.

## Exercício 6 — Buscar apenas 10 produtos

Faça uma requisição `GET` para `/products`, mas retornando apenas 10 produtos.

Use `params` em vez de escrever os parâmetros direto na URL.

In [15]:
url = BASE_URL + "/products"

params = {
    "limit": 10

}

resposta = requests.get(url, params=params)

dados_10 = resposta.json()

print(len(dados_10["products"]))

df_10 = pd.DataFrame(dados_10["products"])

print(df_10)

10
   id                          title  \
0   1  Essence Mascara Lash Princess   
1   2  Eyeshadow Palette with Mirror   
2   3                Powder Canister   
3   4                   Red Lipstick   
4   5                Red Nail Polish   
5   6            Calvin Klein CK One   
6   7        Chanel Coco Noir Eau De   
7   8                   Dior J'adore   
8   9             Dolce Shine Eau de   
9  10             Gucci Bloom Eau de   

                                         description    category   price  \
0  The Essence Mascara Lash Princess is a popular...      beauty    9.99   
1  The Eyeshadow Palette with Mirror offers a ver...      beauty   19.99   
2  The Powder Canister is a finely milled setting...      beauty   14.99   
3  The Red Lipstick is a classic and bold choice ...      beauty   12.99   
4  The Red Nail Polish offers a rich and glossy r...      beauty    8.99   
5  CK One by Calvin Klein is a classic unisex fra...  fragrances   49.99   
6  Coco Noir by Chanel i

## Exercício 7 — Paginação com skip

Agora busque os próximos 10 produtos.

Use:

- `limit=10`
- `skip=10`

Depois, compare os IDs dos produtos do Exercício 6 com os IDs deste exercício.

A ideia é perceber que a API está retornando uma segunda página de resultados.

In [16]:
url = BASE_URL + "/products"

params = {
    "limit": 10,
    "skip": 10
}

resposta = requests.get(url, params=params)

dados_pagina_2 = resposta.json()

df_pagina_2 = pd.DataFrame(dados_pagina_2["products"])

print(df_pagina_2[["id", "title", "price"]])

   id                                      title    price
0  11                       Annibale Colombo Bed  1899.99
1  12                      Annibale Colombo Sofa  2499.99
2  13               Bedside Table African Cherry   299.99
3  14  Knoll Saarinen Executive Conference Chair   499.99
4  15           Wooden Bathroom Sink With Mirror   799.99
5  16                                      Apple     1.99
6  17                                 Beef Steak    12.99
7  18                                   Cat Food     8.99
8  19                               Chicken Meat     9.99
9  20                                Cooking Oil     4.99


**Pergunta:**

Qual foi a diferença entre `limit` e `skip`?

**Resposta:**
limit define quantos registros retornar.
skip define quantos registros serão ignorados antes de retornar os resultados.


# Parte 6 — Selecionando campos específicos

Às vezes, uma API retorna muitos campos, mas você precisa apenas de alguns.

A DummyJSON permite usar o parâmetro `select` para escolher quais campos devem retornar.

Exemplo:

```text
select=title,price,category
```

## Exercício 8 — Buscar produtos com campos selecionados

Faça uma requisição para buscar 15 produtos retornando apenas os campos:

- `title`
- `price`
- `category`
- `rating`

Depois, transforme o resultado em DataFrame.

In [17]:
url = BASE_URL + "/products"

params = {
    "limit": 15,
    "select": "title,price,category,rating"
}

resposta = requests.get(url, params=params)

dados_selecionados = resposta.json()

df_selecionados = pd.DataFrame(dados_selecionados["products"])

print(df_selecionados)

    id                                      title    price    category  rating
0    1              Essence Mascara Lash Princess     9.99      beauty    2.56
1    2              Eyeshadow Palette with Mirror    19.99      beauty    2.86
2    3                            Powder Canister    14.99      beauty    4.64
3    4                               Red Lipstick    12.99      beauty    4.36
4    5                            Red Nail Polish     8.99      beauty    4.32
5    6                        Calvin Klein CK One    49.99  fragrances    4.37
6    7                    Chanel Coco Noir Eau De   129.99  fragrances    4.26
7    8                               Dior J'adore    89.99  fragrances    3.80
8    9                         Dolce Shine Eau de    69.99  fragrances    3.96
9   10                         Gucci Bloom Eau de    79.99  fragrances    2.74
10  11                       Annibale Colombo Bed  1899.99   furniture    4.77
11  12                      Annibale Colombo Sofa  2

# Parte 7 — Busca textual na API

Algumas APIs permitem busca por texto.

Na DummyJSON, podemos pesquisar produtos usando:

```text
/products/search?q=phone
```

O parâmetro `q` representa o termo pesquisado.

## Exercício 9 — Pesquisar produtos por palavra-chave

Faça uma busca por produtos relacionados à palavra:

```text
phone
```

Depois:

1. Exiba a quantidade de resultados encontrados;
2. Crie um DataFrame;
3. Exiba as colunas `id`, `title`, `category`, `price` e `rating`.

In [18]:
url = BASE_URL + "/products/search"

params = {
    "q": "phone"
}

resposta = requests.get(url, params=params)

dados_busca = resposta.json()

print("Resultados:", dados_busca["total"])

df_busca = pd.DataFrame(dados_busca["products"])

print(df_busca[["id", "title", "category", "price", "rating"]])

Resultados: 23
     id                                      title            category  \
0   101                   Apple AirPods Max Silver  mobile-accessories   
1   104                       Apple iPhone Charger  mobile-accessories   
2   105                 Apple MagSafe Battery Pack  mobile-accessories   
3   107              Beats Flex Wireless Earphones  mobile-accessories   
4   108  iPhone 12 Silicone Case with MagSafe Plum  mobile-accessories   
5   110                    Selfie Lamp with iPhone  mobile-accessories   
6   111                       Selfie Stick Monopod  mobile-accessories   
7   121                                  iPhone 5s         smartphones   
8   122                                   iPhone 6         smartphones   
9   123                              iPhone 13 Pro         smartphones   
10  124                                   iPhone X         smartphones   
11  125                                   Oppo A57         smartphones   
12  126                

## Exercício 10 — Repetir a busca com outro termo

Agora escolha outro termo de busca.

Sugestões:

- `laptop`
- `watch`
- `bag`
- `shoes`
- `perfume`

Depois, compare os resultados com a busca anterior.

In [19]:
termo = "laptop"

url = BASE_URL + "/products/search"

params = {
    "q": termo
}

resposta = requests.get(url, params=params)

dados_busca_2 = resposta.json()

df_busca_2 = pd.DataFrame(dados_busca_2["products"])

print(df_busca_2[["id", "title", "category", "price", "rating"]])

   id                                 title category    price  rating
0  78  Apple MacBook Pro 14 Inch Space Grey  laptops  1999.99    3.65
1  79   Asus Zenbook Pro Dual Screen Laptop  laptops  1799.99    3.95
2  80                 Huawei Matebook X Pro  laptops  1399.99    4.98
3  81                       Lenovo Yoga 920  laptops  1099.99    2.86
4  82           New DELL XPS 13 9300 Laptop  laptops  1499.99    2.67


**Perguntas:**

1. Qual termo você pesquisou?
2. Quantos resultados foram encontrados?
3. Os produtos encontrados fazem sentido para o termo pesquisado?
4. Que problema poderia acontecer se o usuário digitasse uma palavra errada?

**Respostas:**

1.laptop
2. 4
3. sim
4. nenhum

# Parte 8 — Categorias

Muitas APIs organizam recursos por categoria.

Na DummyJSON, podemos listar categorias de produtos usando:

```text
/products/category-list
```

E buscar produtos de uma categoria específica usando:

```text
/products/category/NOME_DA_CATEGORIA
```

## Exercício 11 — Listar categorias

Faça uma requisição para listar todas as categorias de produtos.

Depois, exiba a lista de categorias.

In [20]:
url = BASE_URL + "/products/category-list"

resposta = requests.get(url)

categorias = resposta.json()

print(categorias)

['beauty', 'fragrances', 'furniture', 'groceries', 'home-decoration', 'kitchen-accessories', 'laptops', 'mens-shirts', 'mens-shoes', 'mens-watches', 'mobile-accessories', 'motorcycle', 'skin-care', 'smartphones', 'sports-accessories', 'sunglasses', 'tablets', 'tops', 'vehicle', 'womens-bags', 'womens-dresses', 'womens-jewellery', 'womens-shoes', 'womens-watches']


## Exercício 12 — Buscar produtos de uma categoria

Escolha uma categoria da lista anterior.

Depois, busque os produtos dessa categoria.

Exemplo de endpoint:

```text
/products/category/smartphones
```

In [21]:
categoria_escolhida = "smartphones"

url = BASE_URL + f"/products/category/{categoria_escolhida}"

resposta = requests.get(url)

dados_categoria = resposta.json()

df_categoria = pd.DataFrame(dados_categoria["products"])

print(
    df_categoria[
        ["id", "title", "category", "price", "rating", "stock"]
    ]
)

     id               title     category    price  rating  stock
0   121           iPhone 5s  smartphones   199.99    2.83     25
1   122            iPhone 6  smartphones   299.99    3.41     60
2   123       iPhone 13 Pro  smartphones  1099.99    4.12     56
3   124            iPhone X  smartphones   899.99    2.51     37
4   125            Oppo A57  smartphones   249.99    3.94     19
5   126   Oppo F19 Pro Plus  smartphones   399.99    3.51     78
6   127             Oppo K1  smartphones   299.99    4.25     55
7   128          Realme C35  smartphones   149.99    4.20     48
8   129            Realme X  smartphones   299.99    3.70     12
9   130           Realme XT  smartphones   349.99    4.58     80
10  131   Samsung Galaxy S7  smartphones   299.99    3.30     67
11  132   Samsung Galaxy S8  smartphones   499.99    4.40      0
12  133  Samsung Galaxy S10  smartphones   699.99    3.06     19
13  134             Vivo S1  smartphones   249.99    3.50     50
14  135             Vivo 

**Perguntas:**

1. Qual categoria você escolheu?
2. Quantos produtos vieram nessa categoria?
3. Qual foi o produto mais caro?
4. Qual foi o produto com maior avaliação?

**Respostas:**

1. smartphones
2. 15
3. iphone 13
4. realme xt

# Parte 9 — Análise de dados com Pandas

Agora vamos usar os dados da API para fazer pequenas análises.

Essa etapa aproxima o consumo de API do trabalho de Ciência de Dados e IA.

## Exercício 13 — Estatísticas dos preços

Usando o DataFrame `df_categoria` ou outro DataFrame de produtos, calcule:

1. Preço médio;
2. Menor preço;
3. Maior preço;
4. Mediana dos preços.

In [22]:
#print("Preço médio:", df_categoria["price"].mean())

print("Menor preço:", df_categoria["price"].min())

print("Maior preço:", df_categoria["price"].max())

print("Mediana:", df_categoria["price"].median())

Menor preço: 149.99
Maior preço: 1099.99
Mediana: 299.99


## Exercício 14 — Produtos mais bem avaliados

Ordene os produtos por `rating`, do maior para o menor.

Depois, exiba os 5 produtos com maior avaliação.

In [24]:
df_ordenado_rating = df_categoria.sort_values(
    by="rating",
    ascending=False
)

print(
    df_ordenado_rating[
        ["title", "category", "price", "rating"]
    ].head()
)

                title     category   price  rating
9           Realme XT  smartphones  349.99    4.58
11  Samsung Galaxy S8  smartphones  499.99    4.40
15           Vivo X21  smartphones  499.99    4.26
6             Oppo K1  smartphones  299.99    4.25
7          Realme C35  smartphones  149.99    4.20


## Exercício 15 — Produtos com estoque baixo

Filtre os produtos com estoque menor ou igual a 10.

Depois, exiba:

- Nome do produto;
- Categoria;
- Preço;
- Estoque.

In [25]:
df_estoque_baixo = df_categoria[
    df_categoria["stock"] <= 10
]

print(
    df_estoque_baixo[
        ["title", "category", "price", "stock"]
    ]
)

                title     category   price  stock
11  Samsung Galaxy S8  smartphones  499.99      0
15           Vivo X21  smartphones  499.99      7


# Parte 10 — Criando dados com POST

O método `POST` é usado, em geral, para criar um novo recurso.

Na DummyJSON, a criação é simulada.

Endpoint:

```text
POST https://dummyjson.com/products/add
```

Você deverá enviar um corpo JSON com os dados do produto.

## Exercício 16 — Criar um produto fictício

Crie um produto fictício relacionado a tecnologia ou inteligência artificial.

O produto deve ter pelo menos:

- `title`
- `description`
- `price`
- `brand`
- `category`

Depois, envie esse produto usando `POST`.

In [26]:
url = BASE_URL + "/products/add"

novo_produto = {
    "title": "AI Assistant Pro",
    "description": "Assistente inteligente para produtividade",
    "price": 999,
    "brand": "OpenAI Tech",
    "category": "technology"
}

resposta = requests.post(
    url,
    json=novo_produto
)

print(resposta.status_code)
print(resposta.json())

201
{'id': 195, 'title': 'AI Assistant Pro', 'price': 999, 'description': 'Assistente inteligente para produtividade', 'brand': 'OpenAI Tech', 'category': 'technology'}


**Perguntas:**

1. Qual status code foi retornado?
2. A API retornou um ID para o produto criado?
3. Esse produto foi realmente salvo no servidor? Explique com suas palavras.

**Respostas:**

1. 201
2. 195
3. Não, apenas uma simulação da API DummyJSON.

# Parte 11 — Atualizando dados com PUT

O método `PUT` é usado para atualizar um recurso existente.

Na prática, muitas APIs usam `PUT` para atualizar um recurso inteiro e `PATCH` para atualizar apenas alguns campos.

Na DummyJSON, a atualização também é simulada.

Endpoint de exemplo:

```text
PUT https://dummyjson.com/products/1
```

## Exercício 17 — Atualizar o título de um produto

Escolha um produto existente, por exemplo o produto de ID 1.

Atualize o título dele para um nome criado por você.

Depois, exiba a resposta da API.

In [27]:
id_produto = 1

url = BASE_URL + f"/products/{id_produto}"

dados_atualizados = {
    "title": "Notebook IA Ultra"
}

resposta = requests.put(
    url,
    json=dados_atualizados
)

print(resposta.status_code)
print(resposta.json())

200
{'id': 1, 'title': 'Notebook IA Ultra', 'price': 9.99, 'discountPercentage': 10.48, 'stock': 99, 'rating': 2.56, 'images': ['https://cdn.dummyjson.com/product-images/beauty/essence-mascara-lash-princess/1.webp'], 'thumbnail': 'https://cdn.dummyjson.com/product-images/beauty/essence-mascara-lash-princess/thumbnail.webp', 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.', 'brand': 'Essence', 'category': 'beauty'}


## Exercício 18 — Atualizar preço e estoque

Agora atualize dois campos do produto:

- `price`
- `stock`

Use o método `PUT` ou `PATCH`.

In [28]:
id_produto = 1

url = BASE_URL + f"/products/{id_produto}"

dados_atualizados = {
    "price": 1200,
    "stock": 50
}

resposta = requests.patch(
    url,
    json=dados_atualizados
)

print(resposta.json())

{'id': 1, 'title': 'Essence Mascara Lash Princess', 'price': 1200, 'discountPercentage': 10.48, 'stock': 50, 'rating': 2.56, 'images': ['https://cdn.dummyjson.com/product-images/beauty/essence-mascara-lash-princess/1.webp'], 'thumbnail': 'https://cdn.dummyjson.com/product-images/beauty/essence-mascara-lash-princess/thumbnail.webp', 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.', 'brand': 'Essence', 'category': 'beauty'}


**Perguntas:**

1. Qual método você usou: PUT ou PATCH?
2. Quais campos você alterou?
3. A API retornou os dados modificados?
4. Qual é a diferença conceitual entre atualizar dados e criar dados?

**Respostas:**

1. PATCH
2. Prince e stock
3. Sim
4. Criar: cadastrar um novo produto chamado "AI Assistant Pro".
Atualizar: alterar o preço ou o estoque de um produto já cadastrado.

# Parte 12 — Excluindo dados com DELETE

O método `DELETE` é usado para excluir um recurso.

Na DummyJSON, a exclusão é simulada.

Endpoint de exemplo:

```text
DELETE https://dummyjson.com/products/1
```

## Exercício 19 — Simular exclusão de um produto

Faça uma requisição `DELETE` para excluir o produto de ID 1.

Depois, verifique se a resposta possui os campos:

- `isDeleted`
- `deletedOn`

In [29]:
id_produto = 1

url = BASE_URL + f"/products/{id_produto}"

resposta = requests.delete(url)

dados_delete = resposta.json()

print(resposta.status_code)

print(json.dumps(dados_delete, indent=2))

print(dados_delete.get("isDeleted"))

200
{
  "id": 1,
  "title": "Essence Mascara Lash Princess",
  "description": "The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.",
  "category": "beauty",
  "price": 9.99,
  "discountPercentage": 10.48,
  "rating": 2.56,
  "stock": 99,
  "tags": [
    "beauty",
    "mascara"
  ],
  "brand": "Essence",
  "sku": "BEA-ESS-ESS-001",
  "weight": 4,
  "dimensions": {
    "width": 15.14,
    "height": 13.08,
    "depth": 22.99
  },
  "warrantyInformation": "1 week warranty",
  "shippingInformation": "Ships in 3-5 business days",
  "availabilityStatus": "In Stock",
  "reviews": [
    {
      "rating": 3,
      "comment": "Would not recommend!",
      "date": "2025-04-30T09:41:02.053Z",
      "reviewerName": "Eleanor Collins",
      "reviewerEmail": "eleanor.collins@x.dummyjson.com"
    },
    {
      "rating": 4,
      "comment": "Very satisfied!",
      "date": "2025

**Perguntas:**

1. O campo `isDeleted` apareceu?
2. O campo `deletedOn` apareceu?
3. Em uma API real, por que o método DELETE precisa ser usado com cuidado?
4. Que tipo de permissão uma API real provavelmente exigiria antes de permitir exclusões?

**Respostas:**

1.
2.
3.
4.

# Parte 13 — Tratamento básico de erros

Nem toda requisição a uma API dá certo.

Podem ocorrer problemas como:

- Endpoint errado;
- Produto inexistente;
- Falta de internet;
- Falta de autorização;
- Erro interno no servidor.

Por isso, é importante verificar o `status_code` antes de confiar nos dados.

## Exercício 20 — Testar um produto inexistente

Faça uma requisição para buscar um produto com ID muito alto, por exemplo:

```text
/products/999999
```

Depois, exiba:

- O status code;
- A resposta em JSON;
- Uma mensagem amigável dizendo se a requisição deu certo ou não.

In [30]:
url = BASE_URL + "/products/999999"

resposta = requests.get(url)

print(resposta.status_code)

if resposta.status_code == 200:
    print("Produto encontrado")
else:
    print("Produto não encontrado ou erro na requisição")

print(resposta.json())

404
Produto não encontrado ou erro na requisição
{'message': "Product with id '999999' not found"}


# Parte 14 — Criando uma função para consumir a API

Em vez de repetir código várias vezes, podemos criar funções.

Isso deixa o código mais limpo e reutilizável.

## Exercício 21 — Criar uma função para buscar produto por ID

Crie uma função chamada `buscar_produto_por_id`.

A função deve receber um ID e retornar os dados do produto.

Requisitos:

- Receber o ID como parâmetro;
- Fazer a chamada para `/products/{id}`;
- Verificar se o status code é 200;
- Retornar o JSON se der certo;
- Retornar `None` se não encontrar o produto.

In [32]:
def buscar_produto_por_id(id_produto):

    url = BASE_URL + f"/products/{id_produto}"

    resposta = requests.get(url)

    if resposta.status_code == 200:
        return resposta.json()

    return None


print(buscar_produto_por_id(1))

print(buscar_produto_por_id(999999))

{'id': 1, 'title': 'Essence Mascara Lash Princess', 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.', 'category': 'beauty', 'price': 9.99, 'discountPercentage': 10.48, 'rating': 2.56, 'stock': 99, 'tags': ['beauty', 'mascara'], 'brand': 'Essence', 'sku': 'BEA-ESS-ESS-001', 'weight': 4, 'dimensions': {'width': 15.14, 'height': 13.08, 'depth': 22.99}, 'warrantyInformation': '1 week warranty', 'shippingInformation': 'Ships in 3-5 business days', 'availabilityStatus': 'In Stock', 'reviews': [{'rating': 3, 'comment': 'Would not recommend!', 'date': '2025-04-30T09:41:02.053Z', 'reviewerName': 'Eleanor Collins', 'reviewerEmail': 'eleanor.collins@x.dummyjson.com'}, {'rating': 4, 'comment': 'Very satisfied!', 'date': '2025-04-30T09:41:02.053Z', 'reviewerName': 'Lucas Gordon', 'reviewerEmail': 'lucas.gordon@x.dummyjson.com'}, {'rating': 5, 'comment': '

# Parte 15 — Mini desafio final

Agora você deverá juntar tudo o que praticou.

## Cenário

Imagine que você foi contratado para iniciar a análise de produtos de uma loja virtual.

Seu trabalho é consumir a API da DummyJSON, buscar produtos, transformar em tabela e gerar alguns insights iniciais.

## Exercício 22 — Relatório simples de produtos

Faça as seguintes etapas:

1. Busque todos os produtos disponíveis usando `limit=0`;
2. Transforme os produtos em um DataFrame;
3. Exiba a quantidade total de produtos;
4. Exiba a quantidade de categorias diferentes;
5. Mostre as 10 categorias com mais produtos;
6. Mostre os 10 produtos mais caros;
7. Mostre os 10 produtos com maior avaliação;
8. Calcule o preço médio por categoria;
9. Escolha uma categoria e explique brevemente o que você percebeu nos dados.

### Dica

Para buscar todos os produtos, use:

```python
params = {"limit": 0}
```

In [36]:
url = BASE_URL + "/products"

params = {
    "limit": 0
}

resposta = requests.get(
    url,
    params=params
)

dados_todos = resposta.json()

df_todos = pd.DataFrame(
    dados_todos["products"]
)

print("Total:", len(df_todos))

print(
    "Categorias:",
    df_todos["category"].nunique()
)

print(
    df_todos["category"]
    .value_counts()
    .head(10)
)

print(
    df_todos.sort_values(
        by="price",
        ascending=False
    )[["title", "price"]].head(10)
)

print(
    df_todos.sort_values(
        by="rating",
        ascending=False
    )[["title", "rating"]].head(10)
)

print(
    df_todos.groupby("category")["price"]
    .mean()
    .sort_values(ascending=False)
)

Total: 194
Categorias: 24
category
kitchen-accessories    30
groceries              27
sports-accessories     17
smartphones            16
mobile-accessories     14
mens-watches            6
fragrances              5
beauty                  5
mens-shirts             5
laptops                 5
Name: count, dtype: int64
                       title     price
169          Durango SXT RWD  36999.99
167          Charger SXT RWD  32999.99
170         Pacifica Touring  31999.99
166              300 Touring  28999.99
168     Dodge Hornet GT Plus  24999.99
190  Rolex Cellini Moonphase  15999.99
114             MotoGP CI.H1  14999.99
97    Rolex Submariner Watch  13999.99
95   Rolex Cellini Moonphase  12999.99
191     Rolex Datejust Women  10999.99
                             title  rating
98                Amazon Echo Plus    4.99
79           Huawei Matebook X Pro    4.98
94   Rolex Cellini Date Black Dial    4.97
41                           Water    4.96
32                        Mulberry 

## Conclusão do mini desafio

Escreva um pequeno parágrafo respondendo:

1. Que tipo de dado essa API fornece?
2. Como esses dados poderiam ser usados em um projeto de IA?
3. Que cuidado você teria antes de usar dados de uma API real?
4. Qual foi a parte mais difícil do exercício?

**Resposta:**

> A API DummyJSON fornece dados simulados de produtos, usuários, carrinhos de compras, posts e outros recursos. Esses dados poderiam ser utilizados em projetos de IA para análise de comportamento de usuários, sistemas de recomendação de produtos, previsão de vendas e classificação de informações. Antes de utilizar dados de uma API real, é importante verificar a qualidade, a confiabilidade, a atualização dos dados e garantir o respeito à privacidade e à segurança das informações. A parte mais difícil do exercício foi entender a estrutura dos dados retornados em JSON e navegar pelos diferentes endpoints da API para obter as informações desejadas.



# Desafio extra — Escolha outro recurso da API

A DummyJSON também possui outros recursos além de produtos.

Exemplos:

- `/users`
- `/posts`
- `/comments`
- `/carts`
- `/todos`
- `/recipes`

Escolha um desses recursos na documentação e crie uma nova análise.

## O que entregar

1. Endpoint escolhido;
2. Explicação do que esse endpoint retorna;
3. Requisição GET funcionando;
4. DataFrame criado com os dados;
5. Pelo menos 3 perguntas respondidas com análise dos dados.

Exemplo de perguntas:

- Quantos registros existem?
- Quais são os campos disponíveis?
- Existe algum campo numérico para análise?
- É possível agrupar por alguma categoria?
- Esses dados poderiam alimentar algum modelo de IA?

In [37]:
# Desafio extra

endpoint = "/users"

url = BASE_URL + endpoint

# Requisição GET
resposta = requests.get(url)

# Converter para JSON
dados_extra = resposta.json()

# Investigar a estrutura do JSON
print("Status Code:", resposta.status_code)
print("Chaves do JSON:", dados_extra.keys())
print("Total de registros:", dados_extra["total"])

# Criar DataFrame
df_extra = pd.DataFrame(dados_extra["users"])

print(df_extra.head())

# Análises

# 1. Quantos usuários existem?
print("Quantidade de usuários:", len(df_extra))

# 2. Quais campos estão disponíveis?
print("Colunas:")
print(df_extra.columns)

# 3. Qual a idade média dos usuários?
print("Idade média:", df_extra["age"].mean())

# 4. Quantos usuários existem por gênero?
print(df_extra["gender"].value_counts())

# 5. Usuário mais velho
print("Usuário mais velho:")
print(df_extra.loc[df_extra["age"].idxmax()][["firstName", "lastName", "age"]])

Status Code: 200
Chaves do JSON: dict_keys(['users', 'total', 'skip', 'limit'])
Total de registros: 208
   id firstName  lastName maidenName  age  gender  \
0   1     Emily   Johnson      Smith   29  female   
1   2   Michael  Williams              36    male   
2   3    Sophia     Brown              43  female   
3   4     James     Davis              46    male   
4   5      Emma    Miller    Johnson   31  female   

                              email             phone  username      password  \
0     emily.johnson@x.dummyjson.com  +81 965-431-3024    emilys    emilyspass   
1  michael.williams@x.dummyjson.com  +49 258-627-6644  michaelw  michaelwpass   
2      sophia.brown@x.dummyjson.com  +81 210-652-2785   sophiab   sophiabpass   
3       james.davis@x.dummyjson.com  +49 614-958-9364    jamesd    jamesdpass   
4       emma.miller@x.dummyjson.com  +91 759-776-1614     emmaj     emmajpass   

   ...                                            address         macAddress  \
0  ...  {'

# Checklist de entrega

Antes de finalizar, confira se você conseguiu:

- [ ] Fazer uma requisição GET simples;
- [ ] Interpretar um JSON;
- [ ] Criar um DataFrame com dados da API;
- [ ] Usar parâmetros na URL;
- [ ] Fazer busca textual;
- [ ] Buscar produtos por categoria;
- [ ] Simular um POST;
- [ ] Simular um PUT ou PATCH;
- [ ] Simular um DELETE;
- [ ] Tratar um caso de erro;
- [ ] Criar uma função para buscar dados;
- [ ] Gerar uma pequena análise com Pandas;
- [ ] Explicar como o consumo de API pode ser útil em projetos de dados e IA.